ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time_worker.c

```diff
@@ -0,0 +1,391 @@
+#include <ruby.h>
+#include <ruby/thread.h>
+#include <ruby/thread_native.h>
+#include <ruby/debug.h>
+#include <stdbool.h>
+#include <signal.h>
+#include "helpers.h"
+#include "ruby_helpers.h"
+#include "collectors_cpu_and_wall_time.h"
+#include "private_vm_api_access.h"
+
+// Used to trigger the periodic execution of Collectors::CpuAndWallTime, which implements all of the sampling logic
+// itself; this class only implements the "doing it periodically" part.
+//
+// This file implements the native bits of the Datadog::Profiling::Collectors::CpuAndWallTimeWorker class
+
+// ---
+// Here be dragons: This component is quite fiddly and probably one of the more complex in the profiler as it deals with
+// multiple threads, signal handlers, global state, etc.
+//
+// ## Design notes for this class:
+//
+// ### Constraints
+//
+// Currently, sampling Ruby threads requires calling Ruby VM APIs that are only safe to call while holding on to the
+// global VM lock (and are not async-signal safe -- cannot be called from a signal handler).
+//
+// @ivoanjo: As a note, I don't we should think of this constraint as set in stone. Since can reach into the Ruby
+// internals, we may be able to figure out a way of overcoming it. But it's definitely going to be hard so for now
```

Comment:
I find the risk of having the Ruby internals change in a minor release, then us segfaulting the host app to be pretty major. I understand that sometimes we don't have other choices, but the stakes are pretty high here.

I guess we can always make it safe by checking `#define`s, if available during compilation, but internals are internals: they can change without warning.

Anyway, just venting here.

Reply:
Yes, completely! Perhaps my

> But it's definitely going to be hard

was somewhat understated, but your concern is 100% valid and something that would need to be addressed, and why I'm classifying that work as "being hard".

The intent of that note is just to leave the door open for others reading the code, so that they don't take "must have GVL" as a dogma that cannot be touched, and more as a "constraint given the current version of the code and the Ruby VM".

One other approach I hope to begin to take here is to submit changes to upstream Ruby to enable us to do some of these things. If successful, we could then use a different, faster, more efficient code path on modern Rubies, and use the slower-but-still-safe older approach for older Rubies.

I've already gotten one fix in (https://github.com/ruby/ruby/pull/6114) many more to go! 🚀